app/gui2/src/util/ast/abstract.ts

```diff
@@ -0,0 +1,924 @@
+    arg: NodeChild,
+  ): App {
+    return new App(
+      span,
+      id,
+      func,
+      null,
+      null,
+      null,
+      arg,
+      null,
+      isToken(arg.node) ? Tree.Type.DefaultApp : Tree.Type.App,
+    )
+  }
+
+  static _named(
+    span: Span,
+    id: AstId | undefined,
+    func: AstWithWhitespace,
+    leftParen: TokWithWhitespace | null,
+    name: TokWithWhitespace,
+    equals: TokWithWhitespace | undefined,
+    arg: NodeChild,
+    rightParen: TokWithWhitespace | null,
+  ) {
+    return new App(
+      span,
+      id,
+      func,
+      leftParen,
+      name,
+      equals ?? { node: token('=', Token.Type.Operator) },
+      arg,
+      rightParen,
+      Tree.Type.NamedApp,
+    )
+  }
+
+  *_rawChildren(): IterableIterator<NodeChild> {
+    yield this._func
+    if (this._leftParen) yield this._leftParen
+    if (this._name) yield this._name
+    if (this._equals)
+      yield { whitespace: this._equals.whitespace ?? this._arg.whitespace, node: this._equals.node }
+    yield this._arg
+    if (this._rightParen) yield this._rightParen
+  }
+}
+
+export class UnaryOprApp extends Ast {
+  private _opr: TokWithWhitespace
+  private _arg: AstWithWhitespace | null
+
+  get operator(): Tok {
+    return this._opr.node
+  }
+
+  get argument(): Ast | null {
+    const id = this._arg?.node
+    return id ? getNode(id) : null
+  }
+
+  constructor(
+    span: Span,
+    id: AstId | undefined,
+    opr: TokWithWhitespace,
+    arg: AstWithWhitespace | null,
+  ) {
+    super(span, id, Tree.Type.UnaryOprApp)
+    this._opr = opr
+    this._arg = arg
+  }
+
+  *_rawChildren(): IterableIterator<NodeChild> {
+    yield this._opr
+    if (this._arg) yield this._arg
+  }
+}
+
+export class NegationOprApp extends UnaryOprApp {
+  constructor(
+    span: Span,
+    id: AstId | undefined,
+    opr: TokWithWhitespace,
+    arg: AstWithWhitespace | null,
+  ) {
+    super(span, id, opr, arg)
+  }
+}
+
+export class OprApp extends Ast {
+  protected _lhs: AstWithWhitespace | null
+  protected _opr: NodeChild[]
+  protected _rhs: AstWithWhitespace | null
+
+  get lhs(): Ast | null {
+    return this._lhs ? getNode(this._lhs.node) : null
+  }
+
+  get rhs(): Ast | null {
+    return this._rhs ? getNode(this._rhs.node) : null
+  }
+
+  constructor(
+    span: Span,
+    id: AstId | undefined,
+    lhs: AstWithWhitespace | null,
+    opr: NodeChild[],
+    rhs: AstWithWhitespace | null,
+  ) {
+    super(span, id, Tree.Type.OprApp)
+    this._lhs = lhs
+    this._opr = opr
+    this._rhs = rhs
+  }
+
+  *_rawChildren(): IterableIterator<NodeChild> {
+    if (this._lhs) yield this._lhs
+    for (const opr of this._opr) yield opr
+    if (this._rhs) yield this._rhs
+  }
+}
+
+export class PropertyAccess extends OprApp {
+  constructor(
+    span: Span,
+    id: AstId | undefined,
+    lhs: AstWithWhitespace | null,
+    opr: TokWithWhitespace,
+    rhs: AstWithWhitespace | null,
+  ) {
+    super(span, id, lhs, [opr], rhs)
+  }
+}
```

Comment:
Why both `NegationOprApp` and `PropertyAccess` does not make any checks of opr tokens? I guess negation should only work with `-` and property access with `.`.